crypto/store/loader_file.c

```diff
@@ -209,32 +213,169 @@ static OSSL_STORE_INFO *try_decode_PKCS12(const char *pem_name,
+                /*
+                 * Check variant 2: locale to UTF-8 translation
+                 * Convert the pass phrase from the current application locale
+                 * via wide chars to UTF-8.  This should ensure that we can
+                 * read containers protected with correctly encoded pass
+                 * phrases (i.e. PKCS#12 files from other softwares as well)
+                 */
+                {
+                    wchar_t wpass[PEM_BUFSIZE];
+                    size_t wpass_len;
+                    unsigned char *pass_utf8 = NULL;
+                    size_t utf8_count, i, j;
+
+                    if ((wpass_len = mbstowcs(wpass, (char *)tpass,
+                                              OSSL_NELEM(wpass)))
+                        == (size_t) -1) {
+                        SYSerr(SYS_F_MBSTOWCS, errno);
+                        goto p12_end;
+                    }
+
+                    /*
+                     * We use our own rather than iconv for two reasons:
+                     * 1. We're not sure iconv is available "everywhere"
+                     * 2. We assume that any normalization is done by
+                     *    mbstowcs() and that a naïve conversion to UTF8
+                     *    is therefore harmless.
+                     */
+                    utf8_count = 0;
+                    for (i = 0; i < wpass_len; i++)
+                        utf8_count += UTF8_putc(NULL, 0, wpass[i]);
+
+                    if ((pass_utf8 = OPENSSL_malloc(utf8_count + 1)) == NULL) {
+                        OSSL_STOREerr(OSSL_STORE_F_TRY_DECODE_PKCS12,
+                                      ERR_R_MALLOC_FAILURE);
+                        goto p12_end;
+                    }
+
+                    for (i = 0, j = 0; i < wpass_len; i++)
+                        j += UTF8_putc(&pass_utf8[j], utf8_count - j, wpass[i]);
+                    pass_utf8[j] = '\0';
+
+                    tofree = pass = (char *)pass_utf8;
+#ifdef OPENSSL_DEBUG_KEYGEN
+                    fprintf(stderr, "Trying password converted from locale to UTF-8:\n");
+                    h__dump((unsigned char *)pass, strlen(pass));
+#endif
+                    if (PKCS12_verify_mac(p12, pass, strlen(pass)))
+                        goto p12_parse;
+
+                    OPENSSL_clear_free(pass_utf8, utf8_count);
+                    tofree = pass = NULL;
+                }
+#endif
+                /*
+                 * Check variant 3: Naïve ISO-8859-1 to UTF-8 translation
+                 * This ensures that a byte sequence that was given to OpenSSL
+                 * <1.1.0 will be given to the PKCS#12 in such a manner that it
+                 * will be converted to the exact same BMPString.  This covers
+                 * the corner case when the pass phrase byte sequence might
+                 * otherwise be mistaken for a legitimate UTF-8 string.
+                 */
+                {
```

Comment:
And this one could be conditional on... er ... the locale charset not being UTF-8 *and* the string actually having chars outside the ASCII set?

Reply:
Or not being ISO8859-1. More caffeine required. Either way it's probably harmless as you only get here in the failure case.

Reply:
If it's ASCII, the resulting string will be exactly the same as the original. Like you said, this is the last fallback, the pure ASCII gets caught by the first case anyway, so I see no real reason to make extra checks.

Reply:
Going back a bit...

> The default PEM callback gives exactly what it got from the built-in UI method, i.e. the exact characters it got fed, except on Windows with OPENSSL_WIN32_UTF8 defined, where whatever was given by the user is converted to UTF-8

No. The PEM callback doesn't give you characters. It gives you bytes. Octets. `char`s, if you must. But they are not characters, in this sense. To interpret those bytes as characters you must know which encoding they use. It might be UTF-8. It might be something else. On Windows I believe it'll be UTF-8 if `OPENSSL_WIN32_UTF8` is defined, and will be the ANSI code page _(er, or is it the console code page, which is *different*?)_ if `OPENSSL_WIN32_UTF8` is not defined?